### PR TITLE
[stm32l4] I2C Driver

### PR DESCRIPTION
**Pull request by strongly-typed**
Tested in hardware with STM32L476.

### CODE REVIEW

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+	{
+		starting = transaction->starting();
+		switch (starting.next)
+		{
+			case xpcc::I2c::OperationAfterStart::Read:
+				reading = transaction->reading();
+				nextOperation = static_cast<xpcc::I2c::Operation>(reading.next);
+
+				DEBUG_STREAM("read op: reading=" << reading.length);
+				// stream << "nextOperation = " << nextOperation << xpcc::endl;
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: btw, you can also do:
```cpp
DEBUG_STREAM("nextOperation =" << nextOperation);
```

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+		case AddressingMode::Bit10:
+			I2C{{ id }}->OAR1 = (I2C_OAR1_OA1EN | I2C_OAR1_OA1MODE | own_address);
+			I2C{{ id }}->CR2  = (I2C_CR2_ADD10);
+			break;
+	}
+
+	// Configure Addressing Master mode
+	if (addressingMode == AddressingMode::Bit10) {
+		I2C1->CR2  = (I2C_CR2_ADD10);
+	}
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: Duplicate and wrong code.

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+			transaction->detaching(xpcc::I2c::DetachCause::FailedToAttach);
+		}
+	}
+	return false;
+}
+
+xpcc::I2cMaster::Error
+xpcc::stm32::I2cMaster{{ id }}::getErrorState()
+{
+	return error;
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in:470]: Oh hey, look, we've got our own version of `errno`!1!!

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		// peripheral frequency clock
+		constexpr uint8_t freq = SystemClock::I2c{{ id }} / 1000000;
+
+		// maximum rise time: assuming its linear:
+		// 1000ns @ 100kHz and 300ns @ 400kHz
+		//   => y = x * m + b, with m = -2.3333ns/kHz, b = 1233.3333ns
+		constexpr float max_rise_time = -2.333333f * (float(baudrate) / 1000.f) + 1233.333333f;
+		// calculate trise
+		constexpr float trise_raw = max_rise_time < 0 ? 0 : std::floor(max_rise_time / (1000.f / freq));
+		constexpr uint8_t trise = trise_raw > 62 ? 63 : (trise_raw + 1);
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in]: I'm working on the correct timing calculation.

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in (hunk 1 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: Thanks. That's much nicer!

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in (hunk 2 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: Uups, from copy-paste ...

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in (hunk 3 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in:470]: I did not invent it :-)

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in (hunk 4 above)
Comment (salkinium) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in]: I've looked at the documentation, and this is a bit more complicated than I can currently spend time on, so I suggest the following to not block this PR: Let's keep this configuration hardcoded for now, and replace these computations with a `static_assert` (and `// FIXME:`) for 100kHz baudrate at 48MHz clock.

Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.hpp.in]: No problem. A hardcoded LUT would be OK for the beginning. The LUT now has one entry ;-) Just pushed the change.

**Review by salkinium — APPROVED** [review]
Nice!

File: src/xpcc/architecture/platform/devices/stm32/stm32l476_486-j_m_q_r_v_z-c_e_g.xml [hunk 5]
```diff
@@ -129,7 +129,7 @@
     <driver type="adc" name="stm32l4" instances="1,2,3"/>
     <driver type="can" name="stm32" instances="1"/>
     <driver type="clock" name="stm32"/>
-    <driver type="i2c" name="stm32" instances="1,2,3"/>
+    <driver type="i2c" name="stm32l4" instances="1,2,3"/>
```
Comment (salkinium) [inline src/xpcc/architecture/platform/devices/stm32/stm32l476_486-j_m_q_r_v_z-c_e_g.xml:132]: Is this driver also compatible with the other L4 devices? If so, can you also [change the DFG output](https://github.com/roboterclubaachen/xpcc/blob/develop/tools/device_file_generator/dfg/stm32/stm_writer.py#L77), so I don't forget when porting to modm?

**Review by salkinium — CHANGES_REQUESTED** [review]

File: src/xpcc/architecture/platform/devices/stm32/stm32l476_486-j_m_q_r_v_z-c_e_g.xml (hunk 5 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/devices/stm32/stm32l476_486-j_m_q_r_v_z-c_e_g.xml:132]: By visually comparing the I2C register map from all STM32L series

| Reference Manual | Series | Compatible |
| ---------- | ---------- |---|
| RM0038 | STM32L100xx, STM32L151xx, STM32L152xx and STM32L162xx | NO |
| RM0351 | STM32L4x5 and STM32L4x6 | YES |
| RM0377 | STM32L0x1 | YES |
| RM0392 | STM32L4x1 | YES |
| RM0393 | STM32L4x2 | YES |
| RM0394 | STM32L43xxx STM32L44xxx STM32L45xxx STM32L46xxx | YES |
| RM0395 | STM32L4x5 | YES |


I found no obvious difference for STM32L0 and STM32L4. So it seems to be a general rule that the L0 and L4 series uses this 'new' I2C IP. So the name stm32l4 might not be optimal.

The STM32L1 series looks like the STM32F4 but missing the FLTR register.

I updated the DFG without testing. @salkinium Could you please update the device files?

**Comment by strongly-typed** [comment]
Maybe my last comment here got lost ...

I visually compared the I2C register maps from the reference manuals of all STM32L* processors.

| Reference Manual | MCUs | I2C_CR1 PECEN |
|---|---|---|
| RM0038 | STM32L100xx, STM32L151xx, STM32L152xx and STM32L162xx | missing |
| RM0351 | STM32L4x5 and STM32L4x6 |   |
| RM0377 | STM32L0x1 |   |
| RM0392 | STM32L4x1 |   |
| RM0393 | STM32L4x2 |   |
| RM0394 | STM32L43xxx STM32L44xxx STM32L45xxx STM32L46xxx |   |
| RM0395 | STM32L4x5 | |

STM32L4 and STM32L0 seem to share the same I2C, but STM32L1 does not.

**Comment by salkinium** [comment]
> STM32L4 and STM32L0 seem to share the same I2C, but STM32L1 does not.

Yeah, **obviously**. Why simple, when it can be complicated?2??

**Review by salkinium — APPROVED** [review]

**Comment by salkinium** [comment]
Thanks for the good work!

**Pull request merged**